diff --git a/grade-lab-traps b/grade-lab-traps
index 058e77b..66d7c16 100755
--- a/grade-lab-traps
+++ b/grade-lab-traps
@@ -56,6 +56,10 @@ def test_alarmtest_test1():
 def test_alarmtest_test2():
     r.match('^\\.?test2 passed$')
 
+@test(10, "alarmtest: test3", parent=test_alarmtest)
+def test_alarmtest_test3():
+    r.match('^test3 passed$')
+
 @test(19, "usertests")
 def test_usertests():
     r.run_qemu(shell_script([
diff --git a/user/alarmtest.c b/user/alarmtest.c
index 38f09ff..8c29965 100644
--- a/user/alarmtest.c
+++ b/user/alarmtest.c
@@ -15,6 +15,7 @@
 void test0();
 void test1();
 void test2();
+void test3();
 void periodic();
 void slow_handler();
 
@@ -24,6 +25,7 @@ main(int argc, char *argv[])
   test0();
   test1();
   test2();
+  test3();
   exit(0);
 }
 
@@ -155,3 +157,34 @@ slow_handler()
   sigalarm(0, 0);
   sigreturn();
 }
+
+//
+// dummy alarm handler; after running immediately uninstall
+// itself and finish signal handling
+void dummy()
+{
+  sigalarm(0, 0);
+  sigreturn();
+}
+
+//
+// tests that the return from sys_sigreturn() does not
+// modify the a0 register
+void test3()
+{
+  uint64 a0;
+
+  sigalarm(1, dummy);
+  printf("test3 start\n");
+
+  asm volatile("lui a5, 0");
+  asm volatile("addi a0, a5, 0xef" : : : "a0");
+  for(int i = 0; i < 500000000; i++)
+    ;
+  asm volatile("mv %0, a0" : "=r" (a0) );
+
+  if(a0 != 0xef)
+    printf("test3 failed: register a0 changed\n");
+  else
+    printf("test3 passed\n");
+}
